### PR TITLE
Add support for missing columns to OrcSelectiveRecordReader

### PR DESCRIPTION
**Pull request by oerling**
Adds adaptation to missing columns in OrcSelectiveRecordReader. Evaluates TupleDomainFilters against these and fails the whole split if any filter on a missing column fails with null.
Tests interaction of schema evolution with prefilled (e.g. partitioning) columns.

### CODE REVIEW

**Comment by oerling** [comment]
This change is not intended.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 1]
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
 
         requireNonNull(constantValues, "constantValues is null");
         this.constantValues = new Object[this.hiveColumnIndices.length];
+        for (Integer columnIndex : includedColumns.keySet()) {
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: Integer -> int

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 2]
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
...
+            if (!presentColumns.contains(columnIndex)) {
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: use isColumnPresent method from the base class: if (!isColumnPresent(columnIndex)) {

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 3]
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
...
+                // A Any filter except is null on a missing column
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: typo: "A Any" -> "Any"

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 4]
```diff
@@ -160,6 +164,20 @@ public OrcSelectiveRecordReader(
...
+                // fails the whole split. Filters on prefilled columns
+                // are already valuated, hence we only check filters
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: typo? valuated -> evaluated

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 5]
```diff
@@ -171,6 +189,11 @@ public OrcSelectiveRecordReader(
         streamReaderOrder = orderStreamReaders(columnTypes.keySet().stream().filter(index -> this.constantValues[index] == null).collect(toImmutableSet()), columnsWithFilters, filterFunctionInputs);
     }
 
+    private static boolean containsNonNullFilter(Map<Subfield, TupleDomainFilter> columnFilters)
+    {
+        return columnFilters != null && !columnFilters.entrySet().stream().map(Map.Entry::getValue).allMatch(filter -> filter.testNull());
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java]: - entrySet().stream().map(Map.Entry::getValue) can be simplified to values().stream()
- use TupleDomainFilter::testNull in allMatch
- should we check if filters are deterministic as well?

return columnFilters != null && !columnFilters.values().stream().allMatch(TupleDomainFilter::testNull);

File: presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java [hunk 6]
```diff
@@ -73,7 +73,7 @@
 
     private final long totalRowCount;
     private final long splitLength;
-    private final Set<Integer> presentColumns;
+    protected final Set<Integer> presentColumns;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java]: this change is not needed; use isColumnPresent method instead

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 7]
```diff
@@ -514,6 +514,33 @@ public void testPathColumn()
         assertQuerySucceeds(session, "SELECT linenumber, \"$path\" FROM lineitem");
         assertQuerySucceeds(session, "SELECT linenumber, \"$path\" FROM lineitem WHERE length(\"$path\") % 2 = linenumber % 2");
     }
+    @Test
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: add empty line before @Test

**Comment by oerling** [comment]
I addressed the comments. As for determinism, I don’t think we have nondeterministic filters at the place in question. They are made deeper down. So if you have (map[1] = 2 or map[1] is null), then this is true and map[1] = 1 is false and there is no matter of selective application of filters to some elements and not others. For filter functions a determinism check is definitely indicated but for TupleDomain ones I think we are fine without.

**Comment by oerling** [comment]
Updated. What would you suggest about the removed settings?

**Pull request merged**